Error: panic: send to closed channel

_Originally posted by @jonnylangefeld in https://github.com/Azure/azure-event-hubs-go/pull/137#issuecomment-528029362_

This current release actually does sometimes write to a closed channel now:

```
panic: send on closed channel

goroutine 616518 [running]:
github.com/Azure/azure-event-hubs-go/v2/storage.(*LeaserCheckpointer).GetLeases.func1(0xc0002766c0, 0xc003a5c330, 0xc001374780, 0xc0004631a6, 0x2)
	/go/pkg/mod/github.com/!azure/azure-event-hubs-go/v2@v2.0.2/storage/storage.go:199 +0x15a
created by github.com/Azure/azure-event-hubs-go/v2/storage.(*LeaserCheckpointer).GetLeases
	/go/pkg/mod/github.com/!azure/azure-event-hubs-go/v2@v2.0.2/storage/storage.go:197 +0x258
```

sorry, no solution yet. 
@devigned do you know why that could happen? To me your suggestion with the switch case seemed totally valid. Also [this blog post](https://blog.golang.org/context) sound like `ctx.Error()` will only be called after the empty struct was sent to the done channel. 

I'll run more tests and compare it to the solution with the `if` clause I proposed originally. Feel free to make other suggestions.

--- a/changelog.md
+++ b/changelog.md
@@ -1,6 +1,7 @@
 # Change Log
 
 ## `head`
+- fix send on closed channel for GetLeases
 
 ## `v2.0.2`
 - enable partitionKey for sendBatch to fix [#128](https://github.com/Azure/azure-event-hubs-go/issues/128)

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -191,7 +191,6 @@ func (sl *LeaserCheckpointer) GetLeases(ctx context.Context) ([]eph.LeaseMarker,
 
 	partitionIDs := sl.processor.GetPartitionIDs()
 	leaseCh := make(chan leaseGetResult)
-	defer close(leaseCh)
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
@@ -503,7 +502,6 @@ func (sl *LeaserCheckpointer) persistDirtyPartitions(ctx context.Context) error 
 	defer span.End()
 
 	resCh := make(chan dirtyResult)
-	defer close(resCh)
 
 	// gather all of the dirty partition ids
 	pids := make([]string, len(sl.dirtyPartitions))
@@ -517,9 +515,10 @@ func (sl *LeaserCheckpointer) persistDirtyPartitions(ctx context.Context) error 
 	for _, pid := range pids {
 		go func(id string) {
 			err := sl.persistLease(ctx, id)
-			resCh <- dirtyResult{
-				Err:         err,
-				PartitionID: id,
+			select {
+			case <-ctx.Done():
+				return
+			case resCh <- dirtyResult{PartitionID: id, Err: err}:
 			}
 		}(pid)
 	}

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -192,6 +192,8 @@ func (sl *LeaserCheckpointer) GetLeases(ctx context.Context) ([]eph.LeaseMarker,
 	partitionIDs := sl.processor.GetPartitionIDs()
 	leaseCh := make(chan leaseGetResult)
 	defer close(leaseCh)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
 	for _, partitionID := range partitionIDs {
 		go func(pID string) {

